### PR TITLE
EES-2304 EES-2306  Introduce support for methodology version notes

### PR DESCRIPTION
**Pull request by benoutram**
This PR introduces support for methodology version notes along with a new API to create, update and delete them.

* `POST /api/methodologies/{versionId}/notes`
  * Returns the created note view model resource.
* `PUT /api/methodologies/{versionId}/notes/{id}`
  * Returns the updated note view model resource.
* `DELETE /api/methodologies/{versionId}/notes/{id}`
  * Deletes the note and returns no content.

### Other changes

- Clone notes when creating a new methodology version.
  - Notes are cloned for the new version and are assigned a new Id but retain all other values, i.e. the original content, display date, created and last updated dates, created by and last updated by users id's.

### CODE REVIEW

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20210913161056_EES-2171_AddMethodologyNotes.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.CreateTable(
+                name: "MethodologyNotes",
+                columns: table => new
+                {
+                    Id = table.Column<Guid>(nullable: false),
+                    Created = table.Column<DateTime>(nullable: false),
+                    CreatedById = table.Column<Guid>(nullable: false),
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20210913161056_EES-2171_AddMethodologyNotes.cs:16]: Would keep these columns at the end of the table, next to the `Updated*` columns as they're not as important to see

**Review by ntsim — CHANGES_REQUESTED** [review]

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/Methodologies/MethodologyAmendmentServiceTests.cs [hunk 2]
```diff
@@ -196,7 +216,7 @@ await using(var context = InMemoryApplicationDbContext(contextId))
             return new(
                 contentPersistenceHelper ?? new PersistenceHelper<ContentDbContext>(contentDbContext),
                 userService ?? AlwaysTrueUserService().Object,
-                methodologyService ?? new Mock<IMethodologyService>().Object,
+                methodologyService ?? new Mock<IMethodologyService>(Strict).Object,
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/Methodologies/MethodologyAmendmentServiceTests.cs]: Thanks, I've done that now 🙂

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/Methodologies/MethodologyNoteServiceTests.cs [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+            var methodologyNote = new MethodologyNote
+            {
+                MethodologyVersion = methodologyVersion
+            };
+
+            var contentDbContextId = Guid.NewGuid().ToString();
+
+            await using (var contentDbContext = InMemoryApplicationDbContext(contentDbContextId))
+            {
+                await contentDbContext.MethodologyVersions.AddAsync(methodologyVersion);
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/Methodologies/MethodologyNoteServiceTests.cs]: Thanks, I've tidied that up now 🙂

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/MethodologyNoteService.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            MethodologyNoteAddRequest request)
+        {
+            return _persistenceHelper
+                .CheckEntityExists<MethodologyVersion>(methodologyVersionId)
+                .OnSuccess(_userService.CheckCanUpdateMethodology)
+                .OnSuccess(async methodologyVersion =>
+                {
+                    var addedNote = await _methodologyNoteRepository.AddNote(
+                        methodologyVersion.Id,
+                        _userService.GetUserId(),
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/MethodologyNoteService.cs]: Thanks, I like the named parameters. I've added them in now.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/ViewModels/Methodology/MethodologyNoteAddRequest.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿#nullable enable
+using System;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.ViewModels.Methodology
+{
+    public class MethodologyNoteAddRequest
+    {
+        public string Content { get; set; } = null!;
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/ViewModels/Methodology/MethodologyNoteAddRequest.cs]: Cool, I forgot that about `[Required]` preventing the empty string. I've added that in now and also set the property to `string.Empty` in this class and the `MethodologyNoteUpdateRequest` class.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20210913161056_EES-2171_AddMethodologyNotes.cs (hunk 1 above)
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20210913161056_EES-2171_AddMethodologyNotes.cs]: Thanks, I've recreated the migration now and to change the column ordering.

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/Methodologies/MethodologyNoteServiceTests.cs [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+                    methodologyNoteRepository: methodologyNoteRepository.Object);
+
+                var result = await service.DeleteNote(
+                    methodologyVersion.Id,
+                    methodologyNote.Id);
+
+                VerifyAllMocks(methodologyNoteRepository);
+
+                result.AssertRight();
+            }
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/Methodologies/MethodologyNoteServiceTests.cs:100]: I've replaced the mock with a real repository but left the `MethodologyNoteRepositoryTests`.

`MethodologyNoteRepositoryTests` is testing things like the note is really added/updated/removed in the database and comparing dates and user id's etc.

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/Methodologies/MethodologyNoteServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+            )).Returns(Task.CompletedTask);
+
+            await using (var contentDbContext = InMemoryApplicationDbContext(contentDbContextId))
+            {
+                var service = SetupMethodologyNoteService(contentDbContext: contentDbContext,
+                    methodologyNoteRepository: methodologyNoteRepository.Object);
+
+                var result = await service.DeleteNote(
+                    methodologyVersion.Id,
+                    methodologyNote.Id);
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/Methodologies/MethodologyNoteServiceTests.cs]: Thanks, named parameters added in.

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/Methodologies/MethodologyNoteServiceTests.cs [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+
+            var contentDbContextId = Guid.NewGuid().ToString();
+
+            await using (var contentDbContext = InMemoryApplicationDbContext(contentDbContextId))
+            {
+                await contentDbContext.MethodologyVersions.AddAsync(methodologyVersion);
+                await contentDbContext.SaveChangesAsync();
+            }
+
+            var methodologyNoteRepository = new Mock<IMethodologyNoteRepository>(Strict);
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/Methodologies/MethodologyNoteServiceTests.cs]: I've swapped it out for a real repository now which has simplified things quite a bit (i.e. no more returning expected models from the mock and then comparing view model results with that model).

As per my other comment though I left the `MethodologyNoteRepositoryTests` in as they can check the finer grained details like the dates and the user id's are set and updated correctly on the database record.

This leaves the tests in this service asserting values on the returned view models only.

Let me know what you think to this split.

File: src/GovUk.Education.ExploreEducationStatistics.Content.Model.Tests/MethodologyVersionTests.cs [hunk 9]
```diff
@@ -254,7 +255,24 @@ public void CreateMethodologyAmendment()
...
+                        Content = "Note content",
+                        Created = DateTime.Today.AddDays(-4),
+                        CreatedBy = new User(),
+                        CreatedById = Guid.NewGuid(),
+                        DisplayDate = DateTime.Today.ToUniversalTime(),
+                        Updated = DateTime.Today.AddDays(-3),
+                        UpdatedBy = new User(),
+                        UpdatedById = Guid.NewGuid()
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Content.Model.Tests/MethodologyVersionTests.cs]: I've split this test up now to test  cloning multiple notes and all the fields within. The order here is the same as the order that they are added in. I think it's up to the next piece of work, EES-2304 which is going to add a list of all of them to a view model to order that list by their display date which analysts can alter at any time.

**Review by ntsim — DISMISSED** [review]

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/Methodologies/MethodologyNoteServiceTests.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/Methodologies/MethodologyNoteServiceTests.cs]: Think this is fine, although I'd probably say it'd be good to have super rudimentary checks in the service too just to prove things are being deleted/updated/created. It's fine if we don't do this though.

[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/Methodologies/MethodologyNoteServiceTests.cs]: Thanks, just added these checks in as a separate commit if you could take a look please.

**Review by ntsim — APPROVED** [review]

**Pull request merged**